WordPressKit/HTTPRequestBuilder.swift

```diff
@@ -0,0 +1,140 @@
+import Foundation
+
+final class HTTPRequestBuilder {
+    enum Method: String {
+        case get = "GET"
+        case post = "POST"
+        case put = "PUT"
+        case patch = "PATCH"
+        case delete = "DELETE"
+
+        var allowsHTTPBody: Bool {
+            self == .post || self == .put || self == .patch
+        }
+    }
+    private var urlComponents: URLComponents
+    private var method: Method = .get
+    private var path: String
+    private var headers: [String: String] = [:]
+    private var bodyBuilder: ((inout URLRequest) throws -> Void)?
+
+    init(url: URL) {
+        assert(url.scheme == "http" || url.scheme == "https")
+        assert(url.host != nil)
+
+        urlComponents = URLComponents(url: url, resolvingAgainstBaseURL: true)!
+        path = urlComponents.path
+    }
+
+    func set(method: Method) -> Self {
+        self.method = method
+        return self
+    }
+
+    func set(path: String) -> Self {
+        assert(!path.contains("?") && !path.contains("#"), "Path should not have query or fragment: \(path)")
+
+        self.path = path
+        return self
+    }
```

Comment:
How do you feel about creating a type to encapsulate this?

```swift
extension URL {

  struct Path {

      let value: String

      /// `Path` cannot not contain URL query fragments (? and #)
      init(from string: String) {
	assert(!string.contains("?") && !string.contains("#"), "Path should not have query or fragment: \(path)")
        value = string
      }
```

The end result is the same: The app will crash if the path is invalid, but I feel explicitly wrapping it in a dedicated type will make what to do more obvious to consumers.

Reply:
I see value in this dedicated type, if we have more places that need to refer to a "URL path". But for this one method in `HTTPRequestBuilder`, I don't feel like it's worth the effort?

Reply:
Yes, fair enough.